### PR TITLE
Expose the custom type marshalling types for source-generated interop.

### LINKED ISSUE #46838: Attributes and Analyzer Diagnostics to support custom struct marshalling in the DllImport Source Generator
## Background and Motivation

As part of the [DllImport Source Generator proposal](https://github.com/dotnet/runtime/blob/main/docs/design/features/source-generator-pinvokes.md), we need a mechanism to allow users to enable marshalling their struct types from managed code to native code. This issue proposes [a design](https://github.com/dotnet/runtime/blob/main/docs/design/libraries/DllImportGenerator/StructMarshalling.md) along with [an extension](https://github.com/dotnet/runtime/blob/main/docs/design/libraries/DllImportGenerator/SpanMarshallers.md#design-2-out-line-default-support-with-extensions-to-native-type-marshalling-for-contiguous-collections) we have been using in the [`DllImportGenerator`](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Runtime.InteropServices/gen/DllImportGenerator) that is flexible enough to enable us to create a struct interop source generator if we so desire. This design uses the following attributes as well as a convention-based model for the emitted code as described in the linked design doc. The `DllImportGenerator` will consume the proposed attributes in combination with the [`DisableRuntimeMarshallingAttribute`](https://github.com/dotnet/runtime/issues/60639) for its definition of 'does not require marshalling' to allow marshalling of user-defined types.

## Proposed API

Related proposal: https://github.com/dotnet/runtime/issues/46822

```csharp
namespace System.Runtime.InteropServices
{
    /// <summary>
    /// Indicates the default marshalling when the attributed type is used in source-generated interop.
    /// </summary>
    [AttributeUsage(AttributeTargets.Struct | AttributeTargets.Class)]
    public sealed class NativeMarshallingAttribute : Attribute
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="NativeMarshallingAttribute"/>
        /// </summary>
        /// <param name="nativeType">Type that should be used when marshalling the attributed type. The type must not require marshalling.</param>
        public NativeMarshallingAttribute(Type nativeType) {}
    }

    /// <summary>
    /// Indicates the marshalling to use in source-generated interop.
    /// </summary>
    [AttributeUsage(AttributeTargets.Parameter | AttributeTargets.ReturnValue | AttributeTargets.Field, AllowMultiple = true)]
    public sealed class MarshalUsingAttribute : Attribute
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="MarshalUsingAttribute"/>
        /// </summary>
        public MarshalUsingAttribute() {}

        /// <summary>
        /// Initializes a new instance of the <see cref="MarshalUsingAttribute"/>
        /// </summary>
        /// <param name="nativeType">Type that should be used when marshalling the attributed parameter or field. The type must not require marshalling.</param>
        public MarshalUsingAttribute(Type nativeType) {}

        /// <summary>
        /// Type that should be used when marshalling the attributed parameter, field or, when ElementIndirectionLevel is specified, collection element on the attributed item.
        /// </summary>
        public Type? NativeType { get; }

        /// <summary>
        /// Name of the parameter that contains the number of native collection elements. A value of <see cref="ReturnsCountValue"/> indicates the return value of the p/invoke should be used.
        /// </summary>
        public string CountElementName { get; set; }

        /// <summary>
        /// Number of elements in the native collection.
        /// </summary>
        public int ConstantElementCount { get; set; }

        /// <summary>
        /// The level of indirection in a collection to which this attribute should apply.
        /// </summary>
        public int ElementIndirectionLevel { get; set; }

        /// <summary>
        /// When returned by <see cref="CountElementName" />, indicates the return value of the p/invoke should be used as the count.
        /// </summary>
        public const string ReturnsCountValue = "return-value";
    }

    /// <summary>
    /// Indicates this type is a marshaller type to be used in source-generated interop.
    /// </summary>
    [AttributeUsage(AttributeTargets.Struct)]
    public sealed class CustomTypeMarshallerAttribute : Attribute
    {
        public CustomTypeMarshallerAttribute(Type managedType, CustomTypeMarshallerKind marshallerKind = CustomTypeMarshallerKind.Value)
        {
            ManagedType = managedType;
            MarshallerKind = marshallerKind;
        }

        /// <summary>
       /// The managed type that the attributed type is a marshaller for.
       /// </summary>
        public Type ManagedType { get; }

        /// <summary>
       /// The expected shape of this marshaller type
       /// </summary>
        public CustomTypeMarshallerKind MarshallerKind { get; }

        /// <summary>
        /// The size of the caller-allocated buffer in scenarios where using caller-allocated buffer is support.
        /// </summary>
        public int BufferSize { get; set; }

        /// <summary>
       /// Whether or not the caller-allocated buffer must be stack-allocated.
       /// </summary>
        public bool RequiresStackBuffer { get; set; }

        /// <summary>
        /// This type is used as a placeholder for the first generic parameter when generic parameters cannot be used
        /// to identify the managed type (i.e. when the marshaller type is generic over T and the managed type is T[])
        /// </summary>
        public struct GenericPlaceholder
        {
        }
    }


     /// <summary>
    /// The kind (or shape) of the custom marshaller type.
    /// </summary>
    public enum CustomTypeMarshallerKind
    {
         /// <summary>
        /// This marshaller represents marshalling a single value.
        /// </summary>
        Value,
        /// <summary>
        /// This marshaller represents marshalling a Span-like collection of values
        /// </summary>
        SpanCollection
    }
}
```

The types used with `MarshalUsing`/`NativeMarshalling` or attributed with `GenericContiguousCallectionMarshaller` are expected to conform to a defined shape. Analyzers will be provided to validate their usage.

Naming:

As noted in https://github.com/dotnet/runtime/issues/46838#issuecomment-1024812138, these are three different attributes with three different forms of 'marshal'. We may want to rework the names to all use the same form of 'marshal'.

### `NativeMarshallingAttribute`

The `[NativeMarshalling]` attribute is applied on a struct and points to a struct that does not require marshalling or a struct with a `Value` property that does not required marshalling (as per the StructMarshalling design doc). This attribute indicates the default marshalling for the type when passed to a method that uses source-generated interop such as the DllImportGenerator source generator.

```csharp
[NativeMarshalling(typeof(StringContainerNative))]
public struct StringContainer
{
    public string str1;
    public string str2;
}

[CustomTypeMarshaller(typeof(StringContainer))]
public struct StringContainerNative
{
    public IntPtr str1;
    public IntPtr str2;
    public StringContainerNative(StringContainer managed)
    {
        str1 = Marshal.StringToCoTaskMemUTF8(managed.str1);
        str2 = Marshal.StringToCoTaskMemUTF8(managed.str2);
    }
    public StringContainer ToManaged()
    {
        return new StringContainer
        {
            str1 = Marshal.PtrToStringUTF8(str1),
            str2 = Marshal.PtrToStringUTF8(str2)
        };
    }
    public void FreeNative()
    {
        Marshal.FreeCoTaskMem(str1);
        Marshal.FreeCoTaskMem(str2);
    }
}
```

### `MarshalUsingAttribute`

The `[MarshalUsing]` attribute on a parameter, return type, or field indicates the marshaller type to use for the parameter, return value, or field in this particular case. This choice overrides any default marshalling rules, including the `[NativeMarshalling]` attribute.

```csharp
[GeneratedDllImport("NativeLib", EntryPoint = "get_long_bytes_as_double")]
public static partial double GetLongBytesAsDouble([MarshalUsing(typeof(DoubleToLongMarshaller))] double d);

[CustomTypeMarshaller(typeof(double))]
public struct DoubleToLongMarshaller
{
    public long l;
    public DoubleToLongMarshaller(double d)
    {
        l = MemoryMarshal.Cast<double, long>(MemoryMarshal.CreateSpan(ref d, 1))[0];
    }
    public double ToManaged() => MemoryMarshal.Cast<long, double>(MemoryMarshal.CreateSpan(ref l, 1))[0];
    public long Value
    {
        get => l;
        set => l = value;
    }
}
```

Example of code that could be generated:
```csharp
[System.Runtime.CompilerServices.SkipLocalsInitAttribute]
public static partial double GetLongBytesAsDouble(double d)
{
    long __d_gen_native;
    double __retVal;
    //
    // Setup
    //
    global::SharedTypes.DoubleToLongMarshaller __d_gen_native__marshaller = default;
    //
    // Marshal
    //
    __d_gen_native__marshaller = new(d);
    __d_gen_native = __d_gen_native__marshaller.Value;
    //
    // Invoke
    // 
    __retVal = __PInvoke__(__d_gen_native);
    return __retVal;
    //
    // Local P/Invoke
    //
    [System.Runtime.InteropServices.DllImportAttribute("NativeLib", EntryPoint = "get_long_bytes_as_double")]
    extern static unsafe double __PInvoke__(long d);
}
```

The `CountElementName` and `ConstantElementCount` properties can be used to simply specify collection count information without specifying a native type.

```csharp
[GeneratedDllImport("NativeLib")]
[return: MarshalUsing(ConstantElementCount = 10)]
public static partial byte[] Method(
    [MarshalUsing(CountElementName = "outSize")] out byte[] outBuffer,
    out int outSize);
```

### `SpanCollection` shape

The `[MarshalUsing]` or `[NativeMarshalling]` attribute can be used to point to a type marked `
[CustomTypeMarshaller(typeof(...), CustomTypeMarshallerKind.SpanCollection)]` to support custom marshalling of contiguous collections.

```csharp
[GeneratedDllImport("NativeLib")]
public static partial void Method([MarshalUsing(typeof(ListMarshaller<int>))] List<int> values);

[CustomTypeMarshaller(typeof(List<>), CustomTypeMarshallerKind.SpanCollection)]
public unsafe ref struct ListMarshaller<T>
{
    private List<T> managedList;
    private readonly int sizeOfNativeElement;
    private IntPtr allocatedMemory;

    public ListMarshaller(int sizeOfNativeElement)
        : this(null, sizeOfNativeElement)
    { }

    public ListMarshaller(List<T> managed, int sizeOfNativeElement)
    {
        allocatedMemory = default;
        this.sizeOfNativeElement = sizeOfNativeElement;
        if (managed is null)
        {
            managedList = null;
            NativeValueStorage = default;
            return;
        }

        managedList = managed;
        int spaceToAllocate = Math.Max(managed.Count * sizeOfNativeElement, 1);
        allocatedMemory = Marshal.AllocCoTaskMem(spaceToAllocate);
        NativeValueStorage = new Span<byte>((void*)allocatedMemory, spaceToAllocate);
    }

    public Span<T> ManagedValues => CollectionsMarshal.AsSpan(managedList);

    public Span<byte> NativeValueStorage { get; private set; }

    public void SetUnmarshalledCollectionLength(int length)
    {
        managedList = new List<T>(length);
        for (int i = 0; i < length; i++)
        {
            managedList.Add(default);
        }
    }

    public ref byte GetPinnableReference() => ref NativeValueStorage.GetPinnableReference();
    public byte* Value
    {
        get => (byte*)Unsafe.AsPointer(ref GetPinnableReference());
        set
        {
            if (value is null)
            {
                managedList = null;
                NativeValueStorage = default;
            }
            else
            {
                allocatedMemory = (IntPtr)value;
                NativeValueStorage = new Span<byte>(value, (managedList?.Count ?? 0) * sizeOfNativeElement);
            }
        }
    }

    public void FreeNative() => Marshal.FreeCoTaskMem(allocatedMemory);
}
```

## Analyzer Diagnostics

An analyzer will be provided to validate the usage of these APIs as per the design are defined in https://github.com/dotnet/runtime/blob/29cc2a2c23ce15469575a0b25c8b8b453a1d975f/src/libraries/System.Runtime.InteropServices/gen/DllImportGenerator/Analyzers/ManualTypeMarshallingAnalyzer.cs

These will focus on ensuring the types specified in `MarshalUsingAttribute` or `NativeMarshallingAttribute` and attributed with `CustomTypeMarshallerAttribute` conform to the expected shapes for use with p/invoke source generator. The generator will also error if the types do not conform to the expected shapes:

```csharp
// Shape for Value kind
[CustomTypeMarshaller(typeof(ManagedType))]
public struct Marshaller
{
    // Support for marshalling from managed to native.
    // Required if ToManaged() is not defined.
    // May be omitted if marshalling from managed to native is not required.
    public Marshaller(ManagedType managed) {}

    // Optional.
    // Support for marshalling from managed to native using a caller-allocated buffer
    // Requires the BufferSize field to be set on the CustomTypeMarshaller attribute.
   public Marshaller(ManagedType managed, Span<byte> buffer) {}

    // Support for marshalling from native to managed.
    // Required if constructor taking TManaged is not defined.
    // May be omitted if marshalling from native to managed is not required.
    public ManagedType ToManaged() {}

    // Optional.
    // If defined, Value will be passed to native instead of the Marshaller itself
    // If setter is defined, Value will be set when marshalling from native to managed.
    // NativeType must not require marshalling
    public NativeType Value { get; set; }

    // Optional.
    // If defined and Value is defined, will be called before Value property getter and its return value will be pinned
    public ref NativeType GetPinnableReference() {}

    // Optional.
    // Release native resources.
    public void FreeNative() {}
}

// Shape for SpanCollection kind
[CustomTypeMarshaller(typeof(GenericCollection<>), CustomTypeMarshallerKind.SpanCollection)]
public struct GenericContiguousCollectionMarshallerImpl<T>
{
    // Support for marshalling from native to managed.
    // May be omitted if marshalling from native to managed is not required.
    public GenericContiguousCollectionMarshallerImpl(int nativeSizeOfElement);

    // Support for marshalling from managed to native.
    // May be omitted if marshalling from managed to native is not required.
    public GenericContiguousCollectionMarshallerImpl(GenericCollection<T> collection, int nativeSizeOfElement);

    // Optional.
    // Support for marshalling from managed to native using a caller-allocated buffer
    // Requires the BufferSize field to be set on the CustomTypeMarshaller attribute.
    public GenericContiguousCollectionMarshallerImpl(GenericCollection<T> collection, Span<byte> buffer, int nativeSizeOfElement);

    // Required.
    // Points to the memory where the managed values of the collection are stored (in the marshalling case) or should be stored (in the unmarshalling case).
    public Span<CollectionElement> ManagedValues { get; }

    // Set the expected length of the managed collection based on the parameter/return value/field marshalling information.
    // Required for unmarshalling (native to managed) support.
    public void SetUnmarshalledCollectionLength(int length);

    // Required.
    // Points to the memory where the native values of the collection should be stored.
    public unsafe Span<byte> NativeValueStorage { get; }

    // Required.
    // NativeType must not require marshalling
    public NativeType Value { get; set; }
}
```

## Risks

This proposal relies on a definition of 'does not require marshalling'. The DllImportGenerator will require `DisableRuntimeMarshallingAttribute` (assembly-level) for any user-defined types in p/invokes and emit an error if the attribute is not present. This also means that any usage of the DllImportGenerator will require switching over the entire assembly (rather than individual p/invokes) to use the proposed `MarshalUsing` or `NativeMarshalling` attributes.

The successful usage of these attributes also rely on the user defining types that conform to a certain shape, which can be a bit complicated - especially in the `SpanCollection` case. The UX around writing those types might not be great, as walidation is only provided through analyzer diagnostics and generator errors.

### Fixes

Fixes https://github.com/dotnet/runtime/issues/8719

### LINKED ISSUE #66121: [API Proposal]: Specify supported custom type marshalling features on CustomTypeMarshallerAttribute
### Background and motivation

For our source-generated interop work, we have [recently approved](https://github.com/dotnet/runtime/issues/46838) the `System.Runtime.InteropServices.CustomTypeMarshallerAttribute`, which we will use to mark user-defined custom marshallers to help us guide users to define their marshaller types with the correct shapes (as interfaces are not usable with `ref structs`).

Currently, the design calls for us to emit code fixes to offer up the different members that users may want to add. However, we have no mechanism to allow users to say "I want to support these features in my marshaller, so validate at marshaller authoring time that I have the right members".

This issue proposes an enum and an extension of the `CustomTypeMarshallerAttribute` to enable developers to declaratively state which features they plan to support. This will allow our analyzer to enforce that the members required by each shape for each feature are present, and it will allow users of the marshaller types to know which features they can use when using the marshaller.

Our analyzer will enforce that all members required for the features specified by the users are present. The interop source generator will only use the features of the marshaller type specified in the attribute, even if other members are specified.

### API Proposal

```diff
namespace System.Runtime.InteropServices
{
    public class CustomTypeMarshallerAttribute : Attribute
    {
+        public CustomTypeMarshallerDirection Direction { get; set; } = CustomTypeMarshallerDirection.Ref;
+        public CustomTypeMarshallerFeatures Features { get; set; }
    }

+    [Flags]
+    public enum CustomTypeMarshallerDirection
+    {
+        In = 0x1,
+        Out = 0x2,
+        Ref = CustomTypeMarshallerDirection.In | CustomTypeMarshallerDirection.Out
+    }
+    [Flags]
+    public enum CustomTypeMarshallerFeatures
+    {
+        None = 0x0,
+        UnmanagedResources = 0x1,
+        CallerAllocatedBuffer = 0x2,
+        TwoStageMarshalling = 0x4,
+    }
}
```

### API Usage

```C#
[CustomTypeMarshaller(typeof(string), Features = CustomTypeMarshallerFeatures.UnmanagedResources | CustomTypeMarshallerFeatures.CallerAllocatedBuffer | CustomTypeMarshallerFeatures.TwoStageMarshalling, BufferSize = 0x200 )]
public struct Utf16StringMarshaller
{

    // Required for In direction
    public Utf16StringMarshaller(string s);
    // Required for In direction and CallerAllocatedBuffer feature
    public Utf16StringMarshaller(string s, Span<byte> buffer);

   // Required for In direction and TwoStageMarsahalling feature
    public IntPtr ToNativeValue();
   // Required for Out direction and TwoStageMarsahalling feature
    public void FromNativeValue(IntPtr value);

   // Required for Out direction
    public string ToManaged();

    // Required for UnmanagedResources feature
    public void FreeNative();
}

public struct MultiBoolStruct
{
    public bool b1;
    public bool b2;
}

[CustomTypeMarshaller(typeof(MultiBoolStruct), Direction = CustomTypeMarshallerDirection.In )]
public struct MultiBoolStructMarshaller
{
    // Required for In direction
    public MultiBoolStructMarshaller(MultiBoolStruct s);
}
```

### Alternative Designs

We could instead keep the API as-is today and require the generators to inspect the custom marshaller type's API surface as usage time to ensure that it has all of the required members.

### Risks

Enforcement will depend on an analyzer, with can be turned off. However, if someone turns off the analyzer then they'll just end up with a type that can't be used by the source generator due to their own choices, so I don't think this is a large concern.

We have a limited number of bits for features, so we need to make sure that the features are general enough that we won't exceed our limited bitmask size.

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Fixes #46838
Fixes #66121

### CODE REVIEW

**Comment by jkoritzinsky** [comment]
cc: @dotnet/interop-contrib

**Review by AaronRobinsonMSFT — APPROVED** [review]

File: src/libraries/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.cs [hunk 1]
```diff
@@ -327,6 +327,41 @@ public enum CustomQueryInterfaceResult
         NotHandled = 1,
         Failed = 2,
     }
+    [System.AttributeUsageAttribute(System.AttributeTargets.Struct)]
+    public sealed partial class CustomTypeMarshallerAttribute : Attribute
```
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.cs:331]:
```suggestion
    public sealed partial class CustomTypeMarshallerAttribute : System.Attribute
```

File: src/libraries/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.cs [hunk 2]
```diff
@@ -781,6 +816,17 @@ public partial class MarshalDirectiveException : System.SystemException
         public MarshalDirectiveException(string? message) { }
         public MarshalDirectiveException(string? message, System.Exception? inner) { }
     }
+    [System.AttributeUsageAttribute(System.AttributeTargets.Parameter | System.AttributeTargets.ReturnValue, AllowMultiple = true)]
+    public sealed partial class MarshalUsingAttribute : Attribute
```
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.cs:820]:
```suggestion
    public sealed partial class MarshalUsingAttribute : System.Attribute
```

File: src/libraries/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.cs [hunk 3]
```diff
@@ -793,6 +839,12 @@ public static partial class NativeLibrary
         public static bool TryLoad(string libraryPath, out System.IntPtr handle) { throw null; }
         public static bool TryLoad(string libraryName, System.Reflection.Assembly assembly, System.Runtime.InteropServices.DllImportSearchPath? searchPath, out System.IntPtr handle) { throw null; }
     }
+    [System.AttributeUsageAttribute(System.AttributeTargets.Struct | System.AttributeTargets.Class | System.AttributeTargets.Enum | System.AttributeTargets.Delegate)]
+    public sealed partial class NativeMarshallingAttribute : Attribute
+    {
+        public NativeMarshallingAttribute(Type nativeType) { }
+        public Type NativeType { get { throw null; } }
```
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.cs:846]:
```suggestion
    public sealed partial class NativeMarshallingAttribute : System.Attribute
    {
        public NativeMarshallingAttribute(System.Type nativeType) { }
        public System.Type NativeType { get { throw null; } }
```

**Review by elinor-fung — APPROVED** [review]

**Pull request merged**